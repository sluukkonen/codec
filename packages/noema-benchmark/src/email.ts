import * as n from "noema"
import * as z from "zod"
import { Benchmark } from "./index.js"

const suite: Benchmark<string> = {
  name: "email",
  data: "example@example.com",
  codecs: { noema: n.email, zod: z.string().email() },
}

export default suite
