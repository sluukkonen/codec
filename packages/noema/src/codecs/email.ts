import { SimpleCodec } from "../Codec.js"
import { invalidEmail, InvalidEmail, InvalidString } from "../DecodeError.js"
import { refinement } from "./refinement.js"
import { string } from "./string.js"

// https://html.spec.whatwg.org/#e-mail-state-(type=email)
const emailRegexp =
  /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/

export interface EmailCodec
  extends SimpleCodec<string, InvalidString | InvalidEmail> {
  readonly metadata: {
    readonly tag: "email"
    readonly simple: true
  }
}

export const email: EmailCodec = refinement(
  string,
  (s) => emailRegexp.test(s),
  invalidEmail,
  {
    tag: "email",
    simple: true,
  }
)
