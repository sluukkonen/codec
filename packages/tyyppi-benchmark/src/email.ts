import * as t from "tyyppi"
import * as z from "zod"
import { Benchmark } from "./index.js"

const suite: Benchmark<string> = {
  name: "email",
  data: "example@example.com",
  codecs: { tyyppi: t.email, zod: z.string().email() },
}

export default suite
