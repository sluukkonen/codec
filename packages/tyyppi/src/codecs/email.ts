import { SimpleCodec, SimpleMetadata } from "../Codec.js"
import { invalidFormat } from "../errors/index.js"
import { refinement } from "./refinement.js"
import { string } from "./string.js"

// https://html.spec.whatwg.org/#e-mail-state-(type=email)
const emailRegexp =
  /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/

interface EmailMetadata extends SimpleMetadata {
  readonly tag: "email"
}

export type EmailCodec = SimpleCodec<string, EmailMetadata>

export const email: EmailCodec = refinement(
  string,
  (s) => emailRegexp.test(s),
  () => invalidFormat({ format: "email" }),
  { tag: "email" },
)
