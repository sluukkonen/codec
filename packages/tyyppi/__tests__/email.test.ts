import { email } from "../src/codecs/email.js"
import { expectParseFailure, expectParseSuccess } from "./helpers.js"

test("should accept valid email addresses", () => {
  expectParseSuccess(email, "foo@example.com")
})

test("should not accept other strings", () => {
  expectParseFailure(email, "asdf")
})

test("should not accept non-strings", () => {
  expectParseFailure(email, null)
})
