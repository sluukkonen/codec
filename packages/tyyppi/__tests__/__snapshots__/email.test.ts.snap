// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should not accept non-strings 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected a string, received null",
    "path": [],
    "received": "null",
  },
]
`;

exports[`should not accept other strings 1`] = `
[
  {
    "code": "invalid_format",
    "format": "email",
    "message": "Invalid email",
    "path": [],
  },
]
`;
